fix:solve the problem of generating duplicate url encode for presignedUrl

tiswong opened the pull request:
When clicking an object to generate a shared link on the `console`, if the object name contains special characters, calling `url. String()` in the `client.newS3BucketClient` method will encode the characters, and the `*client.makeTargetURL` method in the `minio-go` module will encode again, which will lead to the unavailability of the generated link and report a exception `The specified key does not exist.`
such as:
```xml
<Error>
<Code>NoSuchKey</Code>
<Message>The specified key does not exist.</Message>
<Key>%E6%96%87%E7%A1%95%E8%80%83%E7%A0%94%EF%BC%9A2013%E5%B9%B4%E8%80%83%E7%A0%94%E6%95%B0%E5%AD%A6%E8%87%AA%E6%B5%8B%E9%A2%98%EF%BC%88%E4%B8%80%EF%BC%89.pdf</Key>
<BucketName>mybucket</BucketName>
<Resource>/mybucket/%E6%96%87%E7%A1%95%E8%80%83%E7%A0%94%EF%BC%9A2013%E5%B9%B4%E8%80%83%E7%A0%94%E6%95%B0%E5%AD%A6%E8%87%AA%E6%B5%8B%E9%A2%98%EF%BC%88%E4%B8%80%EF%BC%89.pdf</Resource>
<RequestId>16BCB09B6E9EFAB0</RequestId>
<HostId>99b18f5b-0179-460f-8bed-5a373401f10f</HostId>
</Error>
```

dvaldivia commented:
I think these lines were just changed in #1232 

@Alevsk thoughts?

Alevsk commented:
@tiswong Which version of MInIO/Console are you using? in the past week we changed the way console handles object names with special characters, console will base64 encode the name and send it to the backend, the backend will decode the base64 and send it directly to MInIO via an madmin api call, there's no URL encode anymore.

Can you please test this with the latest version of Console? and its possible with the master branch

tiswong commented:
> 



> @tiswong Which version of MInIO/Console are you using? in the past week we changed the way console handles object names with special characters, console will base64 encode the name and send it to the backend, the backend will decode the base64 and send it directly to MInIO via an madmin api call, there's no URL encode anymore.
> 
> Can you please test this with the latest version of Console? and its possible with the master branch

yes, 

> @tiswong Which version of MInIO/Console are you using? in the past week we changed the way console handles object names with special characters, console will base64 encode the name and send it to the backend, the backend will decode the base64 and send it directly to MInIO via an madmin api call, there's no URL encode anymore.
> 
> Can you please test this with the latest version of Console? and its possible with the master branch



> @tiswong Which version of MInIO/Console are you using? in the past week we changed the way console handles object names with special characters, console will base64 encode the name and send it to the backend, the backend will decode the base64 and send it directly to MInIO via an madmin api call, there's no URL encode anymore.
> 
> Can you please test this with the latest version of Console? and its possible with the master branch



> @tiswong Which version of MInIO/Console are you using? in the past week we changed the way console handles object names with special characters, console will base64 encode the name and send it to the backend, the backend will decode the base64 and send it directly to MInIO via an madmin api call, there's no URL encode anymore.
> 
> Can you please test this with the latest version of Console? and its possible with the master branch

tiswong commented:
> @tiswong Which version of MInIO/Console are you using? in the past week we changed the way console handles object names with special characters, console will base64 encode the name and send it to the backend, the backend will decode the base64 and send it directly to MInIO via an madmin api call, there's no URL encode anymore.
> 
> Can you please test this with the latest version of Console? and its possible with the master branch
Yes, I tested the minio.release.2021-11-24t23-19-33z version and the master version. In the following code, u.string () will URLEncode.


```go
u, err := url.Parse(endpoint)
	if err != nil {
		return nil, fmt.Errorf("the provided endpoint is invalid")
	}
	if strings.TrimSpace(bucketName) != "" {
		u.Path = path.Join(u.Path, bucketName)
	}
	if strings.TrimSpace(prefix) != "" {
		u.Path = path.Join(u.Path, prefix)
	}
	s3Config := newS3Config(u.String(), claims.STSAccessKeyID, claims.STSSecretAccessKey, claims.STSSessionToken, false)
```

tiswong commented:
> > @tiswong Which version of MInIO/Console are you using? in the past week we changed the way console handles object names with special characters, console will base64 encode the name and send it to the backend, the backend will decode the base64 and send it directly to MInIO via an madmin api call, there's no URL encode anymore.
> > Can you please test this with the latest version of Console? and its possible with the master branch
> > Yes, I tested the minio.release.2021-11-24t23-19-33z version and the master version. In the following code, u.string () will URLEncode. @Alevsk 
> 
> ```go
> u, err := url.Parse(endpoint)
> 	if err != nil {
> 		return nil, fmt.Errorf("the provided endpoint is invalid")
> 	}
> 	if strings.TrimSpace(bucketName) != "" {
> 		u.Path = path.Join(u.Path, bucketName)
> 	}
> 	if strings.TrimSpace(prefix) != "" {
> 		u.Path = path.Join(u.Path, prefix)
> 	}
> 	s3Config := newS3Config(u.String(), claims.STSAccessKeyID, claims.STSSecretAccessKey, claims.STSSessionToken, false)
> ```

tiswong commented:
> @tiswong Which version of MInIO/Console are you using? in the past week we changed the way console handles object names with special characters, console will base64 encode the name and send it to the backend, the backend will decode the base64 and send it directly to MInIO via an madmin api call, there's no URL encode anymore.
> 
> Can you please test this with the latest version of Console? and its possible with the master branch
And in code `https://github.com/minio/minio-go/blob/master/api.go` At line 956, URLEncode is executed again

```go
if bucketName != "" {
		// If endpoint supports virtual host style use that always.
		// Currently only S3 and Google Cloud Storage would support
		// virtual host style.
		if isVirtualHostStyle {
			urlStr = scheme + "://" + bucketName + "." + host + "/"
			if objectName != "" {
				urlStr += s3utils.EncodePath(objectName)
			}
		} else {
			// If not fall back to using path style.
			urlStr = urlStr + bucketName + "/"
			if objectName != "" {
				urlStr += s3utils.EncodePath(objectName)
			}
		}
	}
```

Alevsk commented:
> > > @tiswong Which version of MInIO/Console are you using? in the past week we changed the way console handles object names with special characters, console will base64 encode the name and send it to the backend, the backend will decode the base64 and send it directly to MInIO via an madmin api call, there's no URL encode anymore.
> > > Can you please test this with the latest version of Console? and its possible with the master branch
> > > Yes, I tested the minio.release.2021-11-24t23-19-33z version and the master version. In the following code, u.string () will URLEncode. @Alevsk
> > 
> > 
> > ```go
> > u, err := url.Parse(endpoint)
> > 	if err != nil {
> > 		return nil, fmt.Errorf("the provided endpoint is invalid")
> > 	}
> > 	if strings.TrimSpace(bucketName) != "" {
> > 		u.Path = path.Join(u.Path, bucketName)
> > 	}
> > 	if strings.TrimSpace(prefix) != "" {
> > 		u.Path = path.Join(u.Path, prefix)
> > 	}
> > 	s3Config := newS3Config(u.String(), claims.STSAccessKeyID, claims.STSSecretAccessKey, claims.STSSessionToken, false)
> > ```

Can you share an example of a problematic object name? your PR is literally reverting a fix I sent a couple weeks ago https://github.com/minio/console/pull/1232/files

check whats the value of `u.String()` right before calling `newS3Config`

tiswong commented:
> > > > @tiswong Which version of MInIO/Console are you using? in the past week we changed the way console handles object names with special characters, console will base64 encode the name and send it to the backend, the backend will decode the base64 and send it directly to MInIO via an madmin api call, there's no URL encode anymore.
> > > > Can you please test this with the latest version of Console? and its possible with the master branch
> > > > Yes, I tested the minio.release.2021-11-24t23-19-33z version and the master version. In the following code, u.string () will URLEncode. @Alevsk
> > > 
> > > 
> > > ```go
> > > u, err := url.Parse(endpoint)
> > > 	if err != nil {
> > > 		return nil, fmt.Errorf("the provided endpoint is invalid")
> > > 	}
> > > 	if strings.TrimSpace(bucketName) != "" {
> > > 		u.Path = path.Join(u.Path, bucketName)
> > > 	}
> > > 	if strings.TrimSpace(prefix) != "" {
> > > 		u.Path = path.Join(u.Path, prefix)
> > > 	}
> > > 	s3Config := newS3Config(u.String(), claims.STSAccessKeyID, claims.STSSecretAccessKey, claims.STSSessionToken, false)
> > > ```
> 
> @tiswong Can you share an example of a problematic object name? your PR is literally reverting a fix I sent a couple weeks ago https://github.com/minio/console/pull/1232/files
> 
> check whats the value of `u.String()` right before calling `newS3Config`

such as `你好.pdf`

tiswong commented:
> 

Yes, I see your fix. There is no problem transferring from the front to the backend.

> > > > @tiswong Which version of MInIO/Console are you using? in the past week we changed the way console handles object names with special characters, console will base64 encode the name and send it to the backend, the backend will decode the base64 and send it directly to MInIO via an madmin api call, there's no URL encode anymore.
> > > > Can you please test this with the latest version of Console? and its possible with the master branch
> > > > Yes, I tested the minio.release.2021-11-24t23-19-33z version and the master version. In the following code, u.string () will URLEncode. @Alevsk
> > > 
> > > 
> > > ```go
> > > u, err := url.Parse(endpoint)
> > > 	if err != nil {
> > > 		return nil, fmt.Errorf("the provided endpoint is invalid")
> > > 	}
> > > 	if strings.TrimSpace(bucketName) != "" {
> > > 		u.Path = path.Join(u.Path, bucketName)
> > > 	}
> > > 	if strings.TrimSpace(prefix) != "" {
> > > 		u.Path = path.Join(u.Path, prefix)
> > > 	}
> > > 	s3Config := newS3Config(u.String(), claims.STSAccessKeyID, claims.STSSecretAccessKey, claims.STSSessionToken, false)
> > > ```
> 
> @tiswong Can you share an example of a problematic object name? your PR is literally reverting a fix I sent a couple weeks ago https://github.com/minio/console/pull/1232/files
> 
> check whats the value of `u.String()` right before calling `newS3Config`

Yes, I have carefully read the PR submitted by you, and there is no problem when it is transferred from the front end to the back end. I wanted to try to fix this problem from `minio-admin`, but there are too many methods involved in `minio-admin`

dvaldivia commented:
I think is is already fixed, please re-open if you still see the problem
